### PR TITLE
minGW support for TLS example

### PR DESCRIPTION
**Pull request by elms**
Refactor key handling for examples to use keyblobs. This is required for TLS examples to work with some TPMs using Windows TBS as NV storage is not available due to permissions. Also reduces some duplicate code.

### CODE REVIEW

**Comment by elms** [comment]
@dgarske This is still very much a WIP. I did some more reading and thinking and wanted to confirm a a way forward.

NV is not really an option for running examples with the Windows API. If I understand, the TLS examples must be run after the CSR examples and scripts are generated and use then use the key from NV. If it doesn't the key is different and authentication will fail.

I believe the correct way would be to save the context out to a file (instead of NV) and reload it for each test. Is my understanding of the situation correct?

Other items...If you look at this diffs, I'd appreciate your comment on:
 1.  breaking out [common code](https://github.com/wolfSSL/wolfTPM/pull/130/files#diff-b0715f83fef2e6f5a71a8e5327e5e924fc036bb2822c46e6838a48bcc096b5c2)
2. socket support in a [header](https://github.com/wolfSSL/wolfTPM/pull/130/files#diff-e3d039770629673b42f791aa701f3799931e2a7c808079e85746d7e7afd2304c)

**Comment by dgarske** [comment]
@elms , that is correct. Instead of saving to NV use the disk to store the returned key blob. TPM2_Create returns an encrypted blob that only the TPM can decrypt and the it uses TPM2_Load to reload it. The wrapper does both of these (create and load), but internally in the structure the "private" member has the encrypted blob to store on the disk. Then you can just use TPM2_Load to load it.

File: examples/tpm_test.h [hunk 1]
```diff
@@ -506,6 +506,150 @@ static const byte kTestAesCfb128Msg[] = {
     0xe5,0xfb,0xc1,0x19,0x1a,0x0a,0x52,0xef
 };
 
+#include <wolftpm/tpm2_wrap.h>
+
+inline static int getPrimaryStoragekey(WOLFTPM2_DEV* pDev,
```
Comment (dgarske) [inline examples/tpm_test.h:511]: Traditionally its `static inline`, but either probably works.
I'd be fine with creating a `examples/tpm_test.c` and putting these function in a .c with better naming...

**Review by dgarske — CHANGES_REQUESTED** [review]
Changes look good. Perhaps opt-ing for these functions to be in own .c file?

**Review by dgarske — CHANGES_REQUESTED** [review]
Jeff: Mac OS gets these build errors:

```
In file included from examples/tls/tls_client_notpm.c:31:
./examples/tls/tls_common.h:112:14: error: duplicate case value '35'
        case SOCKET_EWOULDBLOCK:
             ^
/usr/local/include/wolfssl/wolfio.h:263:32: note: expanded from macro 'SOCKET_EWOULDBLOCK'
    #define SOCKET_EWOULDBLOCK EWOULDBLOCK
                               ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/errno.h:130:25: note: expanded from macro 'EWOULDBLOCK'
#define EWOULDBLOCK     EAGAIN          /* Operation would block */
                        ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/errno.h:129:25: note: expanded from macro 'EAGAIN'
#define EAGAIN          35              /* Resource temporarily unavailable */
                        ^
./examples/tls/tls_common.h:111:14: note: previous case defined here
        case SOCKET_EAGAIN:
             ^
/usr/local/include/wolfssl/wolfio.h:264:32: note: expanded from macro 'SOCKET_EAGAIN'
    #define SOCKET_EAGAIN      EAGAIN
                               ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/errno.h:129:25: note: expanded from macro 'EAGAIN'
#define EAGAIN          35              /* Resource temporarily unavailable */
                        ^
In file included from examples/tls/tls_client_notpm.c:31:
./examples/tls/tls_common.h:174:14: error: duplicate case value '35'
        case SOCKET_EWOULDBLOCK:
             ^
/usr/local/include/wolfssl/wolfio.h:263:32: note: expanded from macro 'SOCKET_EWOULDBLOCK'
    #define SOCKET_EWOULDBLOCK EWOULDBLOCK
                               ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/errno.h:130:25: note: expanded from macro 'EWOULDBLOCK'
#define EWOULDBLOCK     EAGAIN          /* Operation would block */
                        ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/errno.h:129:25: note: expanded from macro 'EAGAIN'
#define EAGAIN          35              /* Resource temporarily unavailable */
                        ^
./examples/tls/tls_common.h:173:14: note: previous case defined here
        case SOCKET_EAGAIN:
             ^
/usr/local/include/wolfssl/wolfio.h:264:32: note: expanded from macro 'SOCKET_EAGAIN'
    #define SOCKET_EAGAIN      EAGAIN
                               ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/errno.h:129:25: note: expanded from macro 'EAGAIN'
#define EAGAIN          35              /* Resource temporarily unavailable */
                        ^
  CC       examples/csr/csr.o
```

Perhaps add `__MACH__`?

File: examples/tpm_test_keys.c [hunk 2]
```diff
@@ -0,0 +1,392 @@
+#include "tpm_test.h"
```
Comment (dgarske) [inline examples/tpm_test_keys.c:1]: Please add our legal copyright.

File: examples/tpm_test_keys.c [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+#include "tpm_test_keys.h"
+#include <wolftpm/tpm2_wrap.h>
+
+#define RSA_PUB_FILENAME  "rsa_example_pub.raw"
+#define RSA_PRIV_FILENAME "rsa_example_priv.raw"
+#define ECC_PUB_FILENAME  "ecc_example_pub.raw"
+#define ECC_PRIV_FILENAME "ecc_example_priv.raw"
+
+#if !defined(NO_FILESYSTEM) && !defined(NO_WRITE_TEMP_FILES)
+static inline int writeKeyPair(const char* pub_name,
```
Comment (dgarske) [inline examples/tpm_test_keys.c:11]: Let's remove "inline" from these.

File: examples/tpm_test_keys.h [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+#ifndef _TPM_TEST_KEYS_H_
+#define _TPM_TEST_KEYS_H_
+
+#include <wolftpm/tpm2.h>
+#include <wolftpm/tpm2_wrap.h>
+
+int getPrimaryStoragekey(WOLFTPM2_DEV* pDev,
```
Comment (dgarske) [inline examples/tpm_test_keys.h:28]: Please add `WOLFTPM_LOCAL` if they are not public API's.

File: examples/tpm_test_keys.c [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+                               const TPM2B_PRIVATE* priv) {
+    int rc;
+    FILE*  fp = NULL;
+
+    fp = fopen(pub_name, "wb");
+    if (fp) {
+        rc = (int)fwrite(&pub->publicArea, 1, sizeof(TPMT_PUBLIC), fp);
+        fclose(fp);
+        if (rc != sizeof(TPMT_PUBLIC)) {
+            perror("Writing public key");
```
Comment (dgarske) [inline examples/tpm_test_keys.c:23]: Let's avoid the perror API for portability.

File: examples/tpm_test_keys.c [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+}
+
+static inline int readRawKey(const char* filename,
+                             byte* buffer,
+                             UINT16* size)
+{
+    FILE*  file = NULL;
+    int ret;
+
+    file = fopen(filename, "rb");
```
Comment (dgarske) [inline examples/tpm_test_keys.c:48]: We might consider adding an abstraction XFILE and XFOPEN via wolfSSL... For portability.

File: examples/tpm_test_keys.c [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+        XMEMSET(&key->handle.auth, 0, sizeof(TPM2B_AUTH));
+    }
+
+    /* clear auth */
+    XMEMSET(&pDev->session[0].auth, 0, sizeof(pDev->session[0].auth));
+
+    return rc;
+}
+
+static inline int test_CreateAndLoadKey(WOLFTPM2_DEV* dev,
```
Comment (dgarske) [inline examples/tpm_test_keys.c:131]: This code will be obsolete with PR #131 new wrappers.

**Review by dgarske — CHANGES_REQUESTED** [review]
Hey Jeff, excellent work. Can we do a bit of cleanup around the `tryNV`? I'd prefer to not use NV for our examples and rely on using the disk by default. Including the primary storage key on disk.

File: examples/tpm_test_keys.c (hunk 6 above)
Comment (elms) [reply examples/tpm_test_keys.c]: Think I changed all these.

**Comment by elms** [comment]
@dgarske I want to get your feedback since rebasing, but I am tracking down a few more changes that may be necessary from testing. I should add some additional instructions if nothing else. Testing bench and native turned up some other issues, but they may be unrelated.

File: examples/keygen/keygen.c [hunk 8]
```diff
@@ -95,19 +106,48 @@ int TPM2_Keygen_Example(void* userCtx, int argc, char *argv[])
...
+        if (rc != 0) goto exit;
+
+        auth = (byte*)gKeyAuth;
+        authSz = sizeof(gKeyAuth)-1;
     }
+
     printf("Creating new %s key...\n", TPM2_GetAlgName(alg));
     rc = wolfTPM2_CreateKey(&dev, &newKey, &storage.handle,
-        &publicTemplate, (const byte*)gAiKeyAuth, sizeof(gAiKeyAuth)-1);
+                            &publicTemplate, auth, authSz);
```
Comment (dgarske) [inline examples/keygen/keygen.c:150]:
```
examples/keygen/keygen.c:150:52: error: implicit conversion loses integer precision: 'size_t' (aka 'unsigned long') to 'int' [-Werror,-Wshorten-64-to-32]
                            &publicTemplate, auth, authSz);
                                                   ^~~~~~
```

**Review by dgarske — CHANGES_REQUESTED** [review]
Its looking good! 

Many build errors with `--disable-wolfcrypt`. You might need to add some simple STDIO XFILE, XFOPEN wrappers in `wolftpm/tpm_types.h`.

```
/* stdio, default case */
#include <stdio.h>
#define XFILE      FILE*
#define XFOPEN     fopen
#define XFSEEK     fseek
#define XFTELL     ftell
#define XREWIND    rewind
#define XFREAD     fread
#define XFWRITE    fwrite
#define XFCLOSE    fclose
#define XSEEK_END  SEEK_END
#define XBADFILE   NULL
#define XFGETS     fgets
```

I recommend providing an option for user to override themselves at build  time for portability. Add `WOLFTPM_USER_FILESYSTEM` and `NO_FILESYSTEM`


```
./configure --disable-wolfcrypt && make

In file included from examples/tpm_test_keys.c:23:
examples/tpm_test_keys.h:36:29: error: unknown type name 'RsaKey'
                            RsaKey* pWolfRsaKey,
                            ^
examples/tpm_test_keys.c:53:5: error: unknown type name 'XFILE'; did you mean 'FILE'?
    XFILE  fp = NULL;
    ^~~~~
    FILE
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/_stdio.h:157:3: note: 'FILE' declared here
} FILE;
  ^
examples/tpm_test_keys.c:53:12: error: initializing 'FILE' (aka 'struct __sFILE') with an expression of incompatible type 'void *'
    XFILE  fp = NULL;
           ^    ~~~~
examples/tpm_test_keys.c:59:10: error: implicit declaration of function 'XFOPEN' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
    fp = XFOPEN(filename, "rb");
         ^
In file included from examples/keygen/keyload.c:29:
./examples/tpm_test_keys.h:36:29: error: unknown type name 'RsaKey'
                            RsaKey* pWolfRsaKey,
                            ^
1 error generated.
make[1]: *** [examples/keygen/keyload.o] Error 1
make[1]: *** Waiting for unfinished jobs....
examples/tpm_test_keys.c:60:15: error: use of undeclared identifier 'XBADFILE'
    if (fp != XBADFILE) {
              ^
examples/tpm_test_keys.c:61:9: error: implicit declaration of function 'XFSEEK' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        XFSEEK(fp, 0, XSEEK_END);
        ^
examples/tpm_test_keys.c:61:23: error: use of undeclared identifier 'XSEEK_END'
        XFSEEK(fp, 0, XSEEK_END);
                      ^
examples/tpm_test_keys.c:62:18: error: implicit declaration of function 'XFTELL' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        fileSz = XFTELL(fp);
                 ^
examples/tpm_test_keys.c:63:9: error: implicit declaration of function 'XREWIND' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        XREWIND(fp);
        ^
examples/tpm_test_keys.c:70:22: error: implicit declaration of function 'XFREAD' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        bytes_read = XFREAD(&key->pub, 1, sizeof(key->pub), fp);
                     ^
examples/tpm_test_keys.c:94:5: error: implicit declaration of function 'XFCLOSE' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
    XFCLOSE(fp);
    ^
examples/tpm_test_keys.c:157:29: error: unknown type name 'RsaKey'
                            RsaKey* pWolfRsaKey,
                            ^
examples/tpm_test_keys.c:170:10: error: implicit declaration of function 'wc_InitRsaKey_ex' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
    rc = wc_InitRsaKey_ex(pWolfRsaKey, NULL, tpmDevId);
         ^
examples/tpm_test_keys.c:174:10: error: implicit declaration of function 'wolfTPM2_RsaKey_TpmToWolf' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
    rc = wolfTPM2_RsaKey_TpmToWolf(pDev, key, pWolfRsaKey);
         ^
14 errors generated.
make[1]: *** [examples/tpm_test_keys.o] Error 1
In file included from examples/keygen/keygen.c:29:
./examples/tpm_test_keys.h:36:29: error: unknown type name 'RsaKey'
                            RsaKey* pWolfRsaKey,
                            ^
1 error generated.
```

**Review by dgarske — APPROVED** [review]
Thanks Jeff! Excellent work here.

**Review by dgarske — APPROVED** [review]

**Pull request merged**